### PR TITLE
LPC legacy IAP flash driver

### LINKED ISSUE #6040: Implement flash driver for LPC54114
Add a shim over the mcux flashiap driver to adapt it to the zephyr flash driver interface.

### PR DESCRIPTION
**Pull request by agansari**
- soc_flash_lpc: LPC IAP legacy flash driver
 - enable soc_flash_lpc on LPC54xxx
 - updates mpu_test to support for lpc flash driver

Fixes #6040

Depends on https://github.com/zephyrproject-rtos/hal_nxp/pull/75

### CODE REVIEW

File: drivers/flash/soc_flash_lpc.c [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	if (k_sem_take(&priv->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
+
+	key = irq_lock();
+	prepare_erase_write(offset, len, priv->sector_size);
+	page_size = flash_lpc_parameters.write_block_size;
+	start = offset / page_size;
+	stop = (offset+len-1) / page_size;
+    rc = FLASHIAP_ErasePage(start, stop, CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
```
Comment (nvlsianpu) [inline drivers/flash/soc_flash_lpc.c:72]: alignment

**Comment by nvlsianpu** [comment]
Please ensure flash  unrestricted alignment of reads. (see #23628)

**Comment by agansari** [comment]
> Please ensure flash unrestricted alignment of reads. (see #23628)

@nvlsianpu this device can read without alignment restrictions. 
It respects what #23628 documents.
I don't get what I should do with this, i.e. if any action is required from my side.

**Comment by nvlsianpu** [comment]
>@nvlsianpu this device can read without alignment restrictions.
It respects what #23628 documents.
I don't get what I should do with this, i.e. if any action is required from my side.

I just want to be sure that this PR is compatible with mentioned PR. Didn't have time to look into this PR. Thank for confirmation on this, no  further action needed here.

**Comment by agansari** [comment]
This needs to be merged after 2.4 release, adds support for flash controller on LPC54xxx

**Comment by agansari** [comment]
@nvlsianpu @MaureenHelm @galak 
Now that 2.4 is merged, can we have another look at this PR?

**Comment by MaureenHelm** [comment]
> @nvlsianpu @MaureenHelm @galak
> Now that 2.4 is merged, can we have another look at this PR?

Looks fine, but needs rebasing

**Comment by agansari** [comment]
@MaureenHelm rebased and added the missing copyright

**Comment by galak** [comment]
@agansari can you rebase again, looks like we've merged something that created a conflict.

**Comment by agansari** [comment]
> @agansari can you rebase again, looks like we've merged something that created a conflict.

@galak done!

File: drivers/flash/soc_flash_lpc.c [hunk 2]
```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2020 NXP Limited
```
Comment (MaureenHelm) [inline drivers/flash/soc_flash_lpc.c:2]:
```suggestion
 * Copyright (c) 2020 NXP
```

File: dts/arm/nxp/nxp_lpc54xxx.dtsi [hunk 3]
```diff
@@ -58,9 +62,20 @@
...
-			reg = <0 0x40000>;
+		iap: flash-controller@4009c000 {
+			compatible = "nxp,lpc-iap";
+			label = "FLASH_IAP";
+			reg = <0x4009c000 0x18>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+			flash0: flash@0 {
+				compatible = "soc-nv-flash";
+				reg = <0 0x40000>;
```
Comment (MaureenHelm) [inline dts/arm/nxp/nxp_lpc54xxx.dtsi:73]: I'm working to replace memory sizes with DT_SIZE_K macros in #29420. If that PR gets merged first, please update here accordingly

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/flash/soc_flash_lpc.c (hunk 2 above)
Comment (agansari) [reply drivers/flash/soc_flash_lpc.c]: @MaureenHelm done!

File: dts/arm/nxp/nxp_lpc54xxx.dtsi (hunk 3 above)
Comment (agansari) [reply dts/arm/nxp/nxp_lpc54xxx.dtsi]: i've changed them just for the flash node, so you don't have merge conflicts.

Comment (MaureenHelm) [reply dts/arm/nxp/nxp_lpc54xxx.dtsi]: Thanks. The buildkite check is failing because you didn't also add `#include <mem.h>`

Comment (agansari) [reply dts/arm/nxp/nxp_lpc54xxx.dtsi]: I've added `#include <mem.h>`

**Review by MaureenHelm — APPROVED** [review]

File: dts/bindings/mtd/soc-nv-flash.yaml [hunk 4]
```diff
@@ -17,3 +17,8 @@ properties:
      type: int
      description: address alignment required by flash write operations
      required: false
+
+    sector-size:
+     type: int
+     description: the size of one flash sector
```
Comment (mbolivar-nordic) [inline dts/bindings/mtd/soc-nv-flash.yaml:23]: How is this different than erase-block-size?

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I can't find an upstream Linux reference for `sector-size`.

If we're going to add it, the difference between this value and `erase-block-size` should be clearly documented in the binding.

File: dts/bindings/mtd/soc-nv-flash.yaml (hunk 4 above)
Comment (agansari) [reply dts/bindings/mtd/soc-nv-flash.yaml:23]: @mbolivar-nordic it is according to the manual (UM10914 ch 4.3.5 Flash partitions):

> The size of a sector is 32 KB and the size of a page is 256 Byte.

erase-block-size is actually the page size.
Will updating *-block-size to explain they are the page size do it?

Comment (mbolivar-nordic) [reply dts/bindings/mtd/soc-nv-flash.yaml:23]: > @mbolivar-nordic it is according to the manual (UM10914 ch 4.3.5 Flash partitions):

If that's an SoC specific manual, I'm wondering if perhaps we need multiple compatibles for these nodes.

Comment (agansari) [reply dts/bindings/mtd/soc-nv-flash.yaml:23]: sorry @mbolivar-nordic, this is an old pr and just noticed i changes a common definition. I'm refactoring a few files and pushing the changes.

**Review by MaureenHelm — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@mbolivar-nordic please revisit

**Review by nvlsianpu — APPROVED** [review]

File: dts/arm/nxp/nxp_lpc54xxx.dtsi [hunk 5]
```diff
@@ -8,8 +8,19 @@
 #include <dt-bindings/clock/mcux_lpc_syscon_clock.h>
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/i2c/i2c.h>
+#include <mem.h>
 
 / {
+	aliases{
+		gpio-0 = &gpio0;
+		gpio-1 = &gpio1;
+		mailbox-0 = &mailbox0;
```
Comment (mbolivar-nordic) [inline dts/arm/nxp/nxp_lpc54xxx.dtsi:17]: The commit message doesn't say why these are needed and I do not think that they are.

These redundant aliases used to be needed in the legacy devicetree macros that were deprecated in Zephyr 2.3, but we can refer to nodes directly by node label now. The node identifier `DT_NODELABEL(gpio0)` is just as good as `DT_ALIAS(gpio_0)`, so I'd rather not have any more legacy aliases added in.

**Review by mbolivar-nordic — APPROVED** [review]
I'd like to get the aliases removed if we can, but that's not a blocker.

**Pull request merged**